File: PaymentServer/DummySwishRequest.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace PaymentServer
{
	class DummySwishRequest
	{
        float amount;
        string phone_number;

        public static string callback_url;

        public DummySwishRequest(float amount, string phone_number)
        {
            this.amount = amount;
            this.phone_number = phone_number;
        }

        private string CreateMsgString()
        {
            StringBuilder payload_sb = new StringBuilder("", 50);

            payload_sb.Append("{\"callbackUrl\":\"" + callback_url + "\",\"payeeAlias\":\"");
            payload_sb.Append(phone_number);
            payload_sb.Append("\",\"amount\":");
            payload_sb.Append(amount.ToString());
            payload_sb.Append(",\"currency\":\"SEK\"}");

            return payload_sb.ToString();
        }

		public async Task<Result<string>> Send()
		{
            HttpClient client = new HttpClient();
            HttpRequestMessage req_msg = new HttpRequestMessage(HttpMethod.Put, "http://localhost:9001");

            string send_msg = CreateMsgString(); 
            req_msg.Content = new ByteArrayContent(Encoding.ASCII.GetBytes(send_msg));
            string token;
            HttpResponseMessage msg = null;
            try
            {
                msg = await client.SendAsync(req_msg);
                msg.EnsureSuccessStatusCode();

                string msgBody = await msg.Content.ReadAsStringAsync();

                Console.WriteLine("Recieved payment token");

                // json parsing
                JsonDocument jd = JsonDocument.Parse(msgBody);
                token = jd.RootElement.GetProperty("PaymentRequestToken").GetString();
            }
            catch (Exception)
            {
                return Result<string>.Failure(await getErrorMsg(msg));
            }
            return Result<string>.Success(token);
		}

        async Task<string> getErrorMsg(HttpResponseMessage msg)
        {
            if (msg != null)
            {
                HttpStatusCode status_code;
                status_code = msg.StatusCode;

				switch (status_code) 
				{
					case HttpStatusCode.UnprocessableEntity: // 422
						{
							try
							{
								JsonDocument jd = JsonDocument.Parse(await msg.Content.ReadAsStringAsync());
								string m = jd.RootElement.GetProperty("errorCode").GetString();
								m += ": " + jd.RootElement.GetProperty("errorMessage").GetString();
								//Console.WriteLine("swish failed: " + m);
								return m;
							}
							catch (JsonException)
							{
								return "Currupt message";
							}
						}

					case HttpStatusCode.BadRequest:
						return "400 Payment request malformed";
					case HttpStatusCode.Unauthorized:
						return "401 Authentication Problem";
					case HttpStatusCode.Forbidden:
						return "403 payeeAlias not the same as merchant swish number";
					case HttpStatusCode.UnsupportedMediaType:
						return "415 Content is not \"application/json\"";
					case HttpStatusCode.InternalServerError:
						return "500 Internal swish error";
				}
			}

            return "Failed to read message";
        }

	}

    class DummySwishCallback
    {
        JsonDocument document;

        public DummySwishCallback(string callback_msg)
        {
            document = JsonDocument.Parse(callback_msg);
        }

        public string get_status()
        {
            // status can be:
            //  PAID
            //  CANCELED
            //  ERROR
            //  CREATEAD
            return document.RootElement.GetProperty("status").GetString();
        }
    }
}

File: PaymentServerJava/src/paymentserver/DummySwishRequest.java
package paymentserver;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class DummySwishRequest
{
	private float amount;
	private String phone_number;

	public static String callback_url;

	public DummySwishRequest(float amount, String phone_number)
	{
		this.amount = amount;
		this.phone_number = phone_number;
	}

	private String CreateMsgString()
	{
		StringBuilder payload_sb = new StringBuilder(50);

		payload_sb.append("{\"callbackUrl\":\"" + callback_url + "\",\"payeeAlias\":\"");
		payload_sb.append(phone_number);
		payload_sb.append("\",\"amount\":");
		payload_sb.append(String.valueOf(amount));
		payload_sb.append(",\"currency\":\"SEK\"}");

		return payload_sb.toString();
	}

	public final Result<String> Send()
	{
		HttpClient client = HttpClient.newHttpClient();
		HttpRequest req_msg = HttpRequest.newBuilder()
				.uri(URI.create("http://localhost:9001"))
				.header("Content-Type", "application/json")
				.PUT(HttpRequest.BodyPublishers.ofString(CreateMsgString()))
				.build();

		String token;
		HttpResponse<String> response = null;
		try
		{
			response = client.send(req_msg, HttpResponse.BodyHandlers.ofString());


			System.out.println("Received payment token");

			// json parsing
            JSONObject jo = (JSONObject) new JSONParser().parse(response.body());
            token = jo.get("PaymentRequestToken").toString();
		}
		catch (InterruptedException | IOException | ParseException e)
		{
			return Result.Failure(getErrorMsg(response));
		}
		return Result.Success(token);
	}

	private String getErrorMsg(HttpResponse<String> response)
	{
		if (response != null)
		{
			int status_code;
			status_code = response.statusCode();

			switch (status_code)
			{
				case 422: // 422
				{
						try
						{
							JSONObject jo = (JSONObject) new JSONParser().parse(response.body());
							String m = jo.get("errorCode").toString();
							m += jo.get("errorMessage").toString();
							return m;
						}
						catch (ParseException e)
						{
							return "Corrupt message";
						}
				}

				case 400:
					return "400 Payment request malformed";
				case 401:
					return "401 Authentication Problem";
				case 403:
					return "403 payeeAlias not the same as merchant swish number";
				case 415:
					return "415 Content is not \"application/json\"";
				case 500:
					return "500 Internal swish error";
			}
		}

		return "Failed to read message";
	}

}